setup.py

```diff
@@ -129,40 +133,81 @@ def run(self):
-    install_requirements.append(crypto_backend)
+def append_crypto_to_req(req):
+    """Append custom ANSIBLE_CRYPTO_BACKEND to requirements list.
 
-# specify any extra requirements for installation
-extra_requirements = dict()
-extra_requirements_dir = 'packaging/requirements'
-for extra_requirements_filename in os.listdir(extra_requirements_dir):
-    filename_match = re.search(r'^requirements-(\w*).txt$', extra_requirements_filename)
-    if filename_match:
-        with open(os.path.join(extra_requirements_dir, extra_requirements_filename)) as extra_requirements_file:
-            extra_requirements[filename_match.group(1)] = extra_requirements_file.read().splitlines()
+    pycrypto or cryptography. We choose a default but allow the user to
+    override it. This translates into pip install of the sdist deciding what
+    package to install and also the runtime dependencies that pkg_resources
+    knows about.
+    """
+    crypto_backend = os.environ.get('ANSIBLE_CRYPTO_BACKEND', '').strip()
+    if not crypto_backend:
+        return req
+
+    PYCRYPTO_DIST = 'pycrypto'
+    if crypto_backend.strip() == PYCRYPTO_DIST:
+        # Attempt to set version requirements
+        crypto_backend = '%s >= 2.6' % PYCRYPTO_DIST
+
+    CRYPTO_LIBS = PYCRYPTO_DIST, 'cryptography'
+    is_crypto = lambda r: any(_.lower().startswith(c) for c in CRYPTO_LIBS)
+    is_not_crypto = lambda r: not is_crypto(r)
+
+    return list(filter(is_not_crypto, req)) + [crypto_backend]
+
+
+def read_extras():
+    """Specify any extra requirements for installation."""
+    extras = dict()
+    extra_requirements_dir = 'packaging/requirements'
+    for extra_requirements_filename in os.listdir(extra_requirements_dir):
+        filename_match = re.search(r'^requirements-(\w*).txt$', extra_requirements_filename)
+        if not filename_match:
+            continue
+        extra_req_file_path = os.path.join(extra_requirements_dir, extra_requirements_filename)
+        try:
+            extras[filename_match.group(1)] = read_file(extra_req_file_path).splitlines()
+        except RuntimeError:
+            pass
+    return extras
+
+
+def augment_setup_params(params):
+    """Add dynamically calculated setup params to static ones."""
+    _params = params.copy()
+    # Retrieve the long description from the README
+    _params['long_description'] = read_file('README.rst')
+    try:
+        REQUIREMENTS_SRC = 'requirements.txt'
+        _params['install_requires'] = append_crypto_to_req(read_requirements(REQUIREMENTS_SRC))
+    except RuntimeError as exc:
+        print(str(exc), file=sys.stderr)
+        sys.exit(2)
+    _params['extras_require'] = read_extras()
+    return _params
 
-# Retrieve the long description from the README
-with open('README.rst', 'r') as readme_file:
-    longdesc = readme_file.read()
 
-setup(
+static_setup_params = dict(
```

Comment:
Style note, I finally convinced people a year or so ago that ```{}``` is superior to ```dict()```.  So new code should use the literal syntax.

Reply:
I like `{}` more as well. My motivation was to not produce a huge diff, that's why I kept it.